metagraph/src/annotation/taxonomy/taxonomic_db.cpp

```diff
@@ -0,0 +1,315 @@
+            num_taxid_failed += 1;
+            continue;
+        }
+        normalized_taxid[taxid] = num_nodes++;
+        if (taxid == full_parents_list[taxid]) {
+            root_node = normalized_taxid[taxid];
+            continue;
+        }
+        relevant_taxids.push(full_parents_list[taxid]);
+        // Check if taxid is a leaf and used in fasta header.
+        if (reversed_lookup_table.count(taxid)) {
+            lookup_table[reversed_lookup_table.at(taxid)] = normalized_taxid.at(taxid);
+            node_to_acc_version.push_back(reversed_lookup_table.at(taxid));
+        }
+    }
+    if (num_taxid_failed) {
+        logger->warn("Number taxids succeeded {}, failed {}.", num_nodes, num_taxid_failed);
+    }
+
+    tree.resize(num_nodes);
+    for (const auto &it: normalized_taxid) {
+        TaxId taxid = it.first;
+        if (normalized_taxid[taxid] == root_node) {
+            continue;
+        }
+        tree[normalized_taxid[full_parents_list[taxid]]].push_back(
+                normalized_taxid[taxid]);
+    }
+}
+
+void TaxonomyDB::rmq_preprocessing(const std::vector<NormalizedTaxId> &tree_linearization) {
+    uint num_rmq_rows = log2(tree_linearization.size()) + 1;
+    rmq_data.resize(num_rmq_rows);
+    for (uint64_t i = 0; i < num_rmq_rows; ++i) {
+        rmq_data[i].resize(tree_linearization.size());
+    }
+
+    // Copy tree_linearization on rmq[0].
+    for (uint64_t i = 0; i < tree_linearization.size(); ++i) {
+        rmq_data[0][i] = tree_linearization[i];
+    }
+
+    uint delta = 1;
+    for (uint row = 1; row < num_rmq_rows; ++row) {
+        for (uint64_t i = 0; i + delta < tree_linearization.size(); ++i) {
+            if (node_depth[rmq_data[row-1][i]] > node_depth[rmq_data[row-1][i + delta]]) {
+                rmq_data[row][i] = rmq_data[row-1][i];
+            } else {
+                rmq_data[row][i] = rmq_data[row-1][i + delta];
+            }
+        }
+        delta *= 2;
+    }
+
+    precalc_log2.resize(tree_linearization.size());
+    precalc_pow2.push_back(1);
+    for (uint64_t i = 2; i < tree_linearization.size(); ++i) {
+        precalc_log2[i] = 1 + precalc_log2[i/2];
+        if(precalc_log2[i] > precalc_log2[i-1]) {
+            precalc_pow2.push_back(i);
+        }
+    }
+}
+
+void TaxonomyDB::get_input_accessions(const std::string &fasta_headers_filepath,
+                                      tsl::hopscotch_set<AccessionVersion> &input_accessions) {
+    std::ifstream f(fasta_headers_filepath);
+    std::string line;
+
+    while (getline(f, line) ) {
+        std::string fasta_header = utils::split_string(line, "\t")[0];
+        std::string accession_version = utils::split_string(fasta_header, "|")[3];
+        input_accessions.insert(accession_version);
+    }
+    f.close();
+}
+
+// TODO improve this by parsing the compressed ".gz" version (or use https://github.com/pmenzel/taxonomy-tools)
+void TaxonomyDB::read_lookup_table(const std::string &lookup_table_filepath,
```

Comment:
What can be the best way to implement this method for using maybe the compressed version of lookup table?